### PR TITLE
Accept plain list of lumis in the Run object

### LINKED ISSUE #7941: Problems constructing a Run object
Just for tracking what's to be fixed this month.
Run object can receive the lumi arguments in N different data formats and, it is's somehow broken after the events per lumi developments.

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #7941
Got this exception [1] in WorkQueueManager while validating 1.1.4.patch1.

This Run object is somehow tricky with its arguments, besides the run number argument, the others are positional arguments, which means the class constructor sees a tuple of data, which can even be a tuple of a list of tuples :-) (see unit tests)

I'm open for suggestions on a better way or a better performance for fixing it.

[1]
```
2017-06-15 20:53:42,464:140585872336640:INFO:WMBSHelper:"amaltaro_ReReco_Parents_Agent114_Validation_170615_150800_1173" Injecting block /Cosmics/Commissioning2015-PromptReco-v1/RECO#77f9d258-b5d6-11e4-8732-02163e00d625 (89 files) into wmbs
2017-06-15 20:53:42,466:140585872336640:ERROR:WorkQueue:amaltaro_ReReco_Parents_Agent114_Validation_170615_150800_1173, /Cosmics/Commissioning2015-PromptReco-v1/RECO#77f9d258-b5d6-11e4-8732-02163e00d625: 
creating subscription failed in LQ: 
list index out of range
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WorkQueue.py", line 374, in getWork
    dbsBlock)
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WorkQueue.py", line 442, in _wmbsPreparation
    sub, match['NumOfFilesAdded'] = wmbsHelper.createSubscriptionAndAddFiles(block=dbsBlock)
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WMBSHelper.py", line 417, in createSubscriptionAndAddFiles
    addedFiles = self.addFiles(block)
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WMBSHelper.py", line 449, in addFiles
    self._addDBSFileToWMBSFile(dbsFile, block['PhEDExNodeNames'])
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WMBSHelper.py", line 620, in _addDBSFileToWMBSFile
    storageElements, inFileset=False))
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/WorkQueue/WMBSHelper.py", line 640, in _addDBSFileToWMBSFile
    run = Run(lumi['RunNumber'], lumiSecList)
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/DataStructs/Run.py", line 25, in __init__
    self.extendLumis(newLumis)
  File "/data/srv/wmagent/v1.1.4.patch1/sw/slc6_amd64_gcc493/cms/wmagent/1.1.4.patch1/lib/python2.7/site-packages/WMCore/DataStructs/Run.py", line 163, in extendLumis
    self.eventsPerLumi[lumi[0]] = lumi[1]
IndexError: list index out of range
```

### CODE REVIEW

**Comment by amaltaro** [comment]
BTW, I applied this patch to vocms0224 and WorkQueueManager went through the work that was bugging it. So I think it's good.

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
There was a problem with this code (first commit was not considering tuples), which may happen when we construct the Run object by unpacking the list of lumis `(e.g. Run(1, *[(1, None)])`

These are the possible data formats that can reach the `extendLumis` method:
```
()
(1,)
(1, 2, 3, 4, 5)
([1, 2, 3],)
([(1, 11), (2, 22), (3, 33)],)
([(1, 11), (2, 22), (3, 33), (1, 4), (3, 2)],)
([(1, None), (2, None), (3, None)],)
([(1L, None), (2L, None), (3L, None), (2L, None)],)
(1, 2, 3)
((1, None), (2, None), (3, None))
((1L, None), (2L, None), (3L, None), (2L, None))
((1, 11), (2, 22), (3, 33), (1, 4), (3, 2))
```

which should explain all that convoluted type and first element checking. 
Still to be tested, however, If someone feel like improving that code, go ahead and use these unit tests :)

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
I wasn't expecting long numbers on my initial fix. Updated unit tests to check that case too. I hope this was the last surprise using this Run object. So far so good.

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
[Fix #7937 - Accept plain list of lumis in the Run object](https://trello.com/c/OD5n7jYU/41-fix-7937-accept-plain-list-of-lumis-in-the-run-object)

File: src/python/WMCore/DataStructs/Run.py [hunk 1]
```diff
@@ -157,13 +157,20 @@ def extendLumis(self, lumiList):
...
-            elif isinstance(lumi, (list, tuple)):  # Doesn't exist or is 0 or None
-                self.eventsPerLumi[lumi[0]] = lumi[1]
-            else:  # Just given lumis, not events
-                if lumi not in self.eventsPerLumi:  # Don't overwrite existing events
-                    self.eventsPerLumi[lumi] = None
+            if not isinstance(lumi, (list, tuple)):  # comma separated lumi numbers
+                self.eventsPerLumi[lumi] = None
+            else:
+                if isinstance(lumi, list) and not isinstance(lumi[0], tuple):  # then it's a plain list
+                    self.eventsPerLumi = {l: None for l in lumi}
```
Comment (amaltaro) [inline src/python/WMCore/DataStructs/Run.py]: This dictionary comprehension can crash the job during runtime IF the CMSSW uses python2.6. Posting more info outside of the code review.

**Comment by amaltaro** [comment]
As mentioned in the code, dictionary comprehensions can crash the job during runtime if it's using an old CMSSW release which uses python2.6 instead of py2.7. It happens with our ReRecoSkim template. 

The error message looks a bit misleading in the condor.err file though, 
```
INFO:root:Runing SCRAM
INFO:root:Running PRE scripts
INFO:root:RUNNING SCRAM SCRIPTS
INFO:root:    Invoking command: /cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw-patch/CMSSW_5_3_7_patch5/external/slc5_amd64_gcc462/bin/python2.6 -m WMCore.WMRuntime.ScriptInvoke WMTaskSpace.cmsRun1 SetupCMSSWPset

CRITICAL:root:Error running command
/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/bin/python2 -m WMCore.WMRuntime.ScriptInvoke WMTaskSpace.cmsRun1 SetupCMSSWPset

5

        Scram Command Diagnostic:
        Command : scramv1
        Architecture: slc5_amd64_gcc462
        Executed: /cvmfs/cms.cern.ch/slc5_amd64_gcc462/cms/cmssw-patch/CMSSW_5_3_7_patch5/external/slc5_amd64_gcc462/bin/python2.6 -m WMCore.WMRuntime.ScriptInvoke WMTaskSpace.cmsRun1 SetupCMSSWPset

        Exit Status: 5
        Stdout: None
        Stderr: None
INFO:root:Steps.Executors.CMSSW.post called
```

@ticoann @hufnagel do you see something that I'd be missing? I thought CMSSW Setup would be run with COMP python and only the final cmsRun process using the CMSSW python.
I thought the CMSSWSetup would always be done with the COMP python, and only the cmsRun process would be run with the CMSSW python version.

**Comment by amaltaro** [comment]
And the python incompatibility is fixed and it passes all the tests now.
@ticoann feel free to review. I'm going to make the pylint fixes and that's all from my side.

**Comment by amaltaro** [comment]
This LogDB test is flipping on and off, so I guess we're good.

**Review by ticoann — APPROVED** [review]

**Comment by ticoann** [comment]
We need wmagent branch patch for this, right?

**Comment by amaltaro** [comment]
That's right! I see you've already cherry-picked it into 7953

**Comment by ericvaandering** [comment]
@amaltaro you can't use 1L anymore or you break our python 3 compatibility. There is no reason for testing these separately. 

Please remove those tests. Until you do, every PR is going to be flagged as failing the Python3 compatibility.

Amazing, Jenkins goes belly up for 2 days and both you and @hufnagel sneak in python2 only constructs! :-)

**Comment by amaltaro** [comment]
Yes, I saw jenkins were complaining with those long types :(
However, our real use case passes `long` data to the Run constructor (results from mysql/oracle). So my initial fix here was buggy (and I almost didn't spot that :() and we had no unit tests testing the - current - most common usage of Run.

I can remove that in order to keep py3 compatibility, but if we have to change the Run class, we may easily overlook that case and break it again. Having said that :), should I remove those long tests?

**Comment by ericvaandering** [comment]
Yes. I believe python 2.7 guarantees the compatibility

**Comment by amaltaro** [comment]
Provided we don't do something like `isinstance(var, long)` or only `isinstance(var, int)` when a long can be passed too, we should be fine.

**Pull request merged**